middleware/hosts for /etc/hosts parsing

pmoroney opened the pull request:
This adds a middleware to parse and serve hosts files from issue #678 

I pulled the hostsfile.go logic from https://github.com/golang/go/blob/master/src/net/hosts.go
I left the copyright notice in there, but if there is any additional attribution I need to add, please let me know.

I also left the cache for 5 seconds or rescan logic in from net/hosts.go
I figure that is probably good enough, or I can add a duration parameter for customizing the cache time.

I also didn't have the backend config accept a list of zones, but that should be easy enough for me to add if we want it.

Of course any suggestions are appreciated.

pmoroney commented:
Just realized I got PTR handling wrong while looking at the kubernetes middleware, I'll fix that.
I am also looking at adding the Fallthrough support.

I will have some more time this weekend hopefully.

miekg commented:
Np. I'll give this a first glance over today.

miekg commented on middleware/hosts/testdata/case-hosts:
```diff
@@ -0,0 +1,2 @@
+127.0.0.1	PreserveMe	PreserveMe.local
```
Can we put these files into the source and make `Hostsfile` read from the strings? `strings.NewReader` or similar can be used.

We currently don't have any test data artifacts in the coredns code base which is something I like. Look in the file middleware to see how it is done.

pmoroney replied:
I made them stings and separated the parsing of the hosts file from the opening of the file and the caching portion.

miekg commented on middleware/hosts/hosts.go:
```diff
@@ -0,0 +1,104 @@
+		Next middleware.Handler
+		*Hostsfile
+
+		metrics *metrics.Metrics
```
metrics can be removed.

pmoroney replied:
Removed

miekg commented on middleware/hosts/hosts.go:
```diff
@@ -0,0 +1,104 @@
+			r.Ptr = n
+			answers = append(answers, r)
+		}
+	case dns.TypeA:
```
these address cases do a lot of the same work and this can probably be refactored a little.

pmoroney replied:
Working on this now as well as the PTR fixing.

pmoroney commented on middleware.cfg:
```diff
@@ -39,8 +39,9 @@
-200:proxy:proxy
-210:whoami:whoami
-220:erratic:erratic
+200:hosts:hosts
```
I moved this up.

pmoroney commented on middleware/hosts/setup.go:
```diff
@@ -0,0 +1,76 @@
+		return middleware.Error("hosts", err)
+	}
+
+	// If we have enabled prometheus we should add newly discovered zones to it.
```
Removed

pmoroney commented:
I think I addressed most of the issues.
I still need to add the Fallthrough setting, a list of zones in the config, and add comments.
I am hoping I will have some time tomorrow to get to these.

miekg commented:
Looks good, only need a README.md explaining the middleware, see other examples and `middleware.md` for some hand holding.

Another thing: do we need the `file` keyword in the config? Or just `hosts` with an optional file name that defaults to `/etc/hosts`?

pmoroney commented:
something like the following?
hosts [FILE] [ZONES...]

pmoroney commented:
Also right now it defaults to fallthrough to the next middleware. Do we want to make that optional?
And if the question is for a non-A,AAAA,PTR type right now it falls through, should it return a NOERROR with 0 answers for zones that do exist in the hosts file?
And if so, should there be a NextOrNoerror() function?

miekg commented:
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] middleware/ho..." ]
>Also right now it defaults to fallthrough to the next middleware. Do we want to make that optional?

Yes, we've decided this needs to be explicit, by using a 'fallthrough' keyword
in the config.

>And if the question is for a non-A,AAAA,PTR type right now it falls through, should it return a NOERROR with 0 answers for zones that do exist in the hosts file?

In the case of 'no fallthrough'? It should return refused, just let the
middleware return dns.RcodeRefused and CoreDNS will send the error back to the
client.

>And if so, should there be a NextOrNoerror() function?

Your middleware is different in the a /etc/hosts can basically contain any zone,
which means if you didn't find anything:
* with fallthrough; call next middleware, if none I think SERVFAIL is the
  correct response (it might be REFUSED, but I think SERVFAIL is OK as well).
* no fallthrough, could work with SERVFAIL as well, I think.


/Miek

--
Miek Gieben

pmoroney commented:
Ok, I added fallthrough as an explicit setting.
I also updated the README.md and tests accordingly.

Thanks for all of your advice and coaching!
Let me know if there is anything I have missed.

miekg commented on middleware/hosts/setup.go:
```diff
@@ -0,0 +1,88 @@
+	for c.Next() {
+		if c.Val() == "hosts" { // hosts [FILE] [ZONES...]
+			args := c.RemainingArgs()
+			if len(args) == 0 {
```
This check seems wrong as both FILE and ZONES are optional

pmoroney replied:
Yeah, this does look off. I think the cause of your issue above may be related to the changes I made to add the Fallthrough config. I'll look into this tonight and add more tests.

miekg commented on middleware/hosts/setup.go:
```diff
@@ -0,0 +1,88 @@
+				if os.IsNotExist(err) {
+					log.Printf("[WARNING] File does not exist: %s", h.path)
+				} else {
+					return h, c.Errf("Unable to access hosts file '%s': %v", h.path, err)
```
nit: errors should not start with a capital letter.

miekg commented on middleware/hosts/hosts.go:
```diff
@@ -0,0 +1,178 @@
+	}
+
+	if zone != "" {
+		ips := h.LookupStaticHost(zone)
```
Wouldn't it be better to let `LookupStatisHost` returns `net.IP`s? Now we keep on parsing the same string over and over again in `ipv4Filter` and `ipv6Filter`...

miekg replied:
Are even returns multiple sets; v4 and v6

pmoroney replied:
I thought about that initially but had erred on the side of modifying that file as little as possible. But, now that I have modified it quite a bit for the string vs file processing anyways, it makes more sense to modify it to return exactly what we need. I'll refactor this tonight.

miekg commented on middleware/hosts/hosts.go:
```diff
@@ -0,0 +1,178 @@
+	answers := []dns.RR{}
+
+	// Precheck with the origins, i.e. are we allowed to looks here.
+	if h.Origins != nil {
```
Should origins *always* have something?

miekg replied:
I'm also having trouble understanding, why there is precheck? Most of this code is repeated below?

pmoroney replied:
My assumption was if you are specifying the zones in the corefile, and the query is for a zone that is not in that list, then we want to failover to the next middleware even if failover is not specified. Is that assumption correct, or is there another way to handle that case?

Or, as I just realized may happen, does the c.ServerBlockKeys call return a zone regardless, "." perhaps? I'll test that out later tonight.

miekg commented:
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] middleware/ho..." ]
>Yeah, this does look off. I think the cause of your issue above may be related to the changes I made to add the Fallthrough config. I'll look into this tonight and add more tests.

Added an actual Serve test makes sense as well, see TestLookupDNAME in the file
middleware for inspiration.

/Miek

--
Miek Gieben

miekg commented:
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] middleware/ho..." ]
>pmoroney commented on this pull request.
>
>
>
>> +	Origins     []string
>+	Fallthrough bool
>+}
>+
>+// ServeDNS implements the middleware.Handle interface.
>+func (h Hosts) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg) (int, error) {
>+	state := request.Request{W: w, Req: r}
>+	if state.QClass() != dns.ClassINET {
>+		return dns.RcodeServerFailure, middleware.Error(h.Name(), errors.New("can only deal with ClassINET"))
>+	}
>+	qname := state.Name()
>+
>+	answers := []dns.RR{}
>+
>+	// Precheck with the origins, i.e. are we allowed to looks here.
>+	if h.Origins != nil {
>
>My assumption was if you are specifying the zones in the corefile, and the query is for a zone that is not in that list, then we want to failover to the next middleware even if failover is not specified. Is that assumption correct, or is there another way to handle that case?

Yes, that assumption is correct, but each middleware is always responsible for a
set of zones. That set can't be empty.

>Or, as I just realized may happen, does the c.ServerBlockKeys call return a zone regardless, "." perhaps? I'll test that out later tonight.

The `Matches` check you do, check if the incoming query should be handled by
this middleware or directly given to the next middleware. Ideally this check
should only happen once in a middleware.

/Miek

--
Miek Gieben

pmoroney commented:
Just getting started on this, but was that supposed to be a colon as the first character of that corefile?

miekg commented:
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] middleware/ho..." ]
>Just getting started on this, but was that supposed to be a colon as the first character of that corefile?

No a dot: `.`

~~~
.:1053 {
    chaos
    hosts
}
~~~
from the top of my head. I.e. be authoritative for every domain name, this is
needed because most names in a /etc/hosts file are not qualified and basically
top-level domains, e.g. `localhost`.

/Miek

--
Miek Gieben

pmoroney commented:
I think I have it figured out now.
The problem was rooted in Hostsfile using a host normalization that didn't add the final . to dotless hostnames.
In my refactoring to make Hostsfile more efficient, I now use middleware.Name.Normalize() which fixed the problem.

miekg commented:
Just tried it again. Works nice! With and without fallthrough. I'll give it another pass-over.
